### PR TITLE
[Inductor] make `sym_float` and `sym_floor` work with symbolic shapes

### PR DESCRIPTION
**Pull request by nkaretnikov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90626
* __->__ #90121
* #90120

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Codegen (printed from `torch.fx.graph._python_code`):
```
def forward(self, arg0_1):
    sym_size = torch.ops.aten.sym_size(arg0_1, 2)
    mul = sym_size * 2.0
    sym_float = torch.fx.experimental.symbolic_shapes.sym_float(mul);  mul = None
    floor = math_floor(sym_float)
    arange = torch.ops.aten.arange.default(floor, dtype = torch.float32, device = device(type='cuda', index=0), pin_memory = False);  floor = None
    sym_size_1 = torch.ops.aten.sym_size(arg0_1, 3)
    mul_1 = sym_size_1 * 2.0
    sym_float_1 = torch.fx.experimental.symbolic_shapes.sym_float(mul_1);  mul_1 = None
    floor_1 = math_floor(sym_float_1)
    arange_1 = torch.ops.aten.arange.default(floor_1, dtype = torch.float32, device = device(type='cuda', index=0), pin_memory = False);  floor_1 = None
[...]
```

Error (without this patch):
```
  File "torch/fx/experimental/symbolic_shapes.py", line 107, in sym_float
    return float(a)
  File "/lib/python3.10/site-packages/sympy-1.11.1-py3.10.egg/sympy/core/expr.py", line 351, in __float__
    raise TypeError("Cannot convert expression to float")
TypeError: Cannot convert expression to float
```

Fails because a `sympy` expression is being passed to `floor`. This changes things such that the expression is passed further.

Error (with this patch):
```
  File "torch/_inductor/graph.py", line 273, in call_function
    raise LoweringException(e, target, args, kwargs) from e
torch._inductor.exc.LoweringException: RuntimeError: Overloaded torch operator invoked from Python failed to many any schema:
[...]
aten::arange() Expected a value of type 'number' for argument 'end' but instead found type 'Mul'.
[...]
  target: aten.arange.default
  args[0]: 2.0*s4
  kwargs: {'dtype': torch.float32, 'device': device(type='cuda', index=0), 'pin_memory': False}

While executing %arange : [#users=1] = call_function[target=torch.ops.aten.arange.default](args = (%sym_floor,), kwargs = {dtype: torch.float32, device: cuda:0, pin_memory: False})
[...]
```

This fails because `arange` doesn't work with symbolic values. It expects a `Scalar`, but we're passing a symbolic `Mul`. However, we cannot just use `SymInt` or `SymFloat` there. It's a factory op and its output depends on the type of the argument. 

```
In [136]: torch.arange(2.)
Out[136]: tensor([0., 1.])

In [137]: torch.arange(2)
Out[137]: tensor([0, 1])
```

My current plan is to introduce `SymScalar`, wrapping the two mentioned classes, and using that in `arange`. To be done.

Another approach that I've tried is just forcing the computation with `guard_int` right before the `arange` call. But that uncovered a codegen bug (will be reported separately) and probably not something we want to do anyway.

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -3630,6 +3631,16 @@ def sym_numel(a):
...
 
 
+@register_lowering(sym_float)
+def sym_float(a):
+    return a
+
+
+@register_lowering(sym_floor)
+def sym_floor(a):
+    return a
```
Comment (ezyang) [inline torch/_inductor/lowering.py:3641]: this lowering looks wrong if you have a symfloat lol

**Comment by ezyang** [comment]
I don't really see why `sym_floor` has to become a primitive; as the comments say, Sympy is able to work with math.floor automatically with no changes. Your posted backtrace is for `sym_float` so there isn't any evidence there either.

**Comment by ezyang** [comment]
I don't understand why you need SymScalar in Python. Python is dynamically typed; if you have a SymInt, just pass the SymInt around. No need to wrap it again.

**Comment by nkaretnikov** [comment]
@ezyang 

> I don't really see why sym_floor has to become a primitive; as the comments say, Sympy is able to work with math.floor automatically with no changes. Your posted backtrace is for sym_float so there isn't any evidence there either.

It's because with `upsample_bilinear` decomp, you ended up with something like this:

```
math.floor(sympy.Number(2) * sympy.Symbol("sym_size", number=True))
```

If we look at the generated code again:

```
def forward(self, arg0_1):
    sym_size = torch.ops.aten.sym_size(arg0_1, 2)
    mul = sym_size * 2.0
    sym_float = torch.fx.experimental.symbolic_shapes.sym_float(mul);  mul = None
    floor = math_floor(sym_float)
```

With the current lowerings, the `Mul` expr is wrapped into `SymFloat` and is propagated to `floor` where it fails (because you cannot `floor` an expr).

Without adding special handling for `floor`, the code just attempts to `math.floor` and fails.

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (nkaretnikov) [reply torch/_inductor/lowering.py:3641]: Could you explain in more detail what needs to be done instead? I basically copied what `sym_float` does.

Comment (ezyang) [reply torch/_inductor/lowering.py:3641]: If I pass 4.5 into this function, it needs to go to 4, but right now this noops

**Pull request closed without merging**